### PR TITLE
K8s namespaces

### PR DESCRIPTION
**Pull request by hunchback**
Added new k8s/namespace probe:
- creating a Node per each namespace
- creating an edge { "RelationType": "Association" } between namespaces to {pod,networkpolicy} objects

### CODE REVIEW

File: topology/probes/k8s/graph.go [hunk 1]
```diff
@@ -23,35 +23,45 @@
...
 
 const (
-	manager = "k8s"
-	hostID  = ""
+	managerValue = "k8s"
+	hostID       = ""
+)
+
+const (
+	nameField      = "Name"
```
Comment (lebauce) [inline topology/probes/k8s/graph.go]: I don't really the value of constants -especially nameField and extraField - for this

File: tests/k8s_test.go [hunk 2]
```diff
@@ -110,3 +118,7 @@ func TestK8sNetworkPolicyNode(t *testing.T) {
 func TestK8sNodeNode(t *testing.T) {
 	testNodeCreation(t, nil, nil, gremlin.NewValueString("node").Quote(), gremlin.NewValueString(nodeName).Quote())
 }
+
+func TestK8sNamespaceNode(t *testing.T) {
+	testNodeCreation(t, setupNamespace, tearDownNamespace, gremlin.NewValueString("namespace").Quote(), gremlin.NewValueString(namespaceName).Quote())
```
Comment (safchain) [inline tests/k8s_test.go]: I think it is time to add test to check that the links are created properly. By the way it is a bit hard to read what the test is really checking with this level of indirections. The query builder doesn't help. I think it is more convenient to have a complete gremlin expression readable. fmt.Sprintf(`G.V().Has("Manager", "%s", "Type": "%s", "Name": "%s")`).

File: topology/probes/k8s/graph.go (hunk 1 above)
Comment (hunchback) [reply topology/probes/k8s/graph.go]: Replaced:
- `nameField` -> `"Name"`
- `namespaceField` -> `"Namespace"`

I left `extraField` as this value is used as root for probes metadata ... a root unique for each probe ... this is changes to a standard location (shared by all probes such as `Extra`) then I agree that their is not need for the constant.

File: tests/k8s_test.go (hunk 2 above)
Comment (hunchback) [reply tests/k8s_test.go]: Will take this comment into account when extending the k8s tests (this effort can now continue after accepting the first k8s tests PR). In any case will be handled separately to this PR

File: topology/probes/k8s/namespace.go [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+
+		for _, objNode := range p.objectIndexer.Get(ns.GetName()) {
+			p.linkObject(objNode, nsNode)
+		}
+	}
+}
+
+func (p *namespaceProbe) OnUpdate(oldObj, newObj interface{}) {
+	if ns, ok := newObj.(*v1.Namespace); ok {
+		p.lockAll()
```
[the remark this answers is not in the document]
Comment (hunchback) [reply topology/probes/k8s/namespace.go]: lock has been removed

File: topology/probes/k8s/namespace.go [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+		if nsNode := p.graph.GetNode(namespaceUID(ns)); nsNode != nil {
+			addMetadata(p.graph, nsNode, ns)
+			logging.GetLogger().Debugf("Updated %s", dumpNamespace(ns))
+		}
+	}
+}
+
+func (p *namespaceProbe) OnDelete(obj interface{}) {
+	if ns, ok := obj.(*v1.Namespace); ok {
+		p.lockAll()
```
[the remark this answers is not in the document]
Comment (hunchback) [reply topology/probes/k8s/namespace.go]: lock has been removed

**Comment by lebauce** [comment]
run skydive-functional-tests-backend-elasticsearch

File: tests/k8s_test.go (hunk 2 above)
Comment (safchain) [reply tests/k8s_test.go]: ok thx

**Pull request merged**